server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -158,13 +164,32 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
                     throw new DelayRecoveryException("source node does not have the shard listed in its state as allocated on the node");
                 }
                 assert targetShardRouting.initializing() : "expected recovery target to be initializing but was " + targetShardRouting;
+                retentionLeaseRef.set(useRetentionLeases ? shard.getRetentionLeases().get(
```

Comment:
Is there any issue if we use any existing retention leases for the peer recovery purpose? I mean to rely on "Engine#getMinRetainedSeqNo".

Reply:
Yes, the issue is that we do not guarantee that the primary retains every operation required by every retention lease, so we must use `hasCompleteHistoryOperations` to ensure this. The problem is that in a file-based recovery we create a retention lease at the local checkpoint of this shard's safe commit, which may be behind every other lease, so we cannot be certain that every other peer is also able to respect this lease; if this primary were to fail then another primary may be elected without all the history needed for all its leases.

Reply:
See e.g. this comment: https://github.com/elastic/elasticsearch/pull/44853/files#diff-9feb4947ba4df0b486c7de6b9d46502fR215-R219

Reply:
On reflection, it might actually be possible to do this: we'd need to keep the retention lock open until at least having replayed history to the global checkpoint. I'll think about this a bit more.

Reply:
Ok, Yannick and I discussed this and you're right, there's no real drawback to creating the new replica's retention lease according to the current (persisted) global checkpoint rather than the local checkpoint of the safe commit. This means that we've much greater chance that this retention lease is satisfied on every in-sync shard copy, at which point we should not obviously need to call `hasCompleteHistoryOperations` in many cases.

It is, however, not totally watertight, because today we sometimes create unsatisfied leases for BWC reasons. I haven't gone through the details to see if there are other cases too, but given that it's not something we're asserting today I have concerns that it might be possible. Also it's cheap to call `hasCompleteHistoryOperations` and will save us from disaster, so let's keep it in.

I've adjusted the creation logic in 5fb8bda58b7ee62a8865ece9b1f1160142fd65c9.